File: 1886-minimum-limit-of-balls-in-a-bag/1886-minimum-limit-of-balls-in-a-bag.cs
public class Solution {
    public int MinimumSize(int[] nums, int maxOperations) {
        int l = 1, r = nums.Max();

        while (l < r) {
            int m = l + (r - l) / 2, c = 0;
            foreach (int a in nums)
                 c += (a - 1) / m;
            if (c > maxOperations)
                l = m + 1;
            else
                r = m;
        }
        return l;
    }
}

File: 1886-minimum-limit-of-balls-in-a-bag/1886-minimum-limit-of-balls-in-a-bag.java
class Solution {
    public int minimumSize(int[] nums, int maxOperations) {
        int max_val = 0;
        for (int x : nums)
            max_val = Math.max(max_val, x);

        int l = 1, r = max_val;
        while (l <= r) {
            int m = ((r - l) >> 1) + l;
            int cnt = 0;
            for (int x : nums)
                cnt += (x - 1) / m;
            if (cnt > maxOperations)
                l = m + 1;
            else
                r = m - 1;
        }
        return l;
    }
}